src/utils/input-utils.ts

```diff
@@ -10,205 +19,215 @@ import { DropdownChoice, SomeCompanionInputField } from '../../../../instance_sk
-    type: 'dropdown',
-    label: 'Channel Type',
-    id: 'channelType',
-    ...CHOICES.auxChannelTypes
-  },
-  fxChannelTypeDropdown: {
-    type: 'dropdown',
-    label: 'Channel Type',
-    id: 'channelType',
-    ...CHOICES.fxChannelTypes
-  },
-  busNumberField: {
-    type: 'number',
-    label: 'Bus number',
-    id: 'bus',
-    min: 1,
-    max: 10,
-    default: 1
-  },
-  channelNumberField: {
-    type: 'number',
-    label: 'Channel number',
-    id: 'channel',
-    min: 1,
-    max: 24,
-    default: 1
-  },
-  hwChannelNumberField: {
-    type: 'number',
-    label: 'HW Channel number',
-    id: 'hwchannel',
-    min: 1,
-    max: 20,
-    default: 1
-  },
-  muteDropdown: {
-    type: 'dropdown',
-    label: 'Mute',
-    id: 'mute',
-    ...CHOICES.mute
-  },
-  muteGroupDropdown: {
-    type: 'dropdown',
-    label: 'Group',
-    id: 'group',
-    choices: [
-      { id: 1, label: 'MUTE Group 1' },
-      { id: 2, label: 'MUTE Group 2' },
-      { id: 3, label: 'MUTE Group 3' },
-      { id: 4, label: 'MUTE Group 4' },
-      { id: 5, label: 'MUTE Group 5' },
-      { id: 6, label: 'MUTE Group 6' },
-      { id: 'all', label: 'MUTE ALL' },
-      { id: 'fx', label: 'MUTE FX' }
-    ],
-    default: 1
-  },
-  soloDropdown: {
-    type: 'dropdown',
-    label: 'Solo',
-    id: 'solo',
-    ...CHOICES.onofftoggleDropdown
-  },
-  faderValuesSlider: {
-    type: 'number',
-    label: 'Fader Level dB (-100 = -∞)',
-    id: 'value',
-    range: true,
-    required: true,
-    default: 0,
-    step: 0.1,
-    min: -100,
-    max: 10
-  },
-  prepostDropdown: {
-    type: 'dropdown',
-    label: 'PRE/POST',
-    id: 'post',
-    choices: [
-      { id: 0, label: 'PRE' },
-      { id: 1, label: 'POST' },
-      { id: 2, label: 'Toggle' }
-    ],
-    default: 2
-  },
-  easingsDropdown: {
-    type: 'dropdown',
-    label: 'Easing',
-    id: 'easing',
-    choices: [
-      { id: Easings.Linear, label: 'Linear / None' },
-      { id: Easings.EaseIn, label: 'Ease In' },
-      { id: Easings.EaseOut, label: 'Ease Out' },
-      { id: Easings.EaseInOut, label: 'Ease In Out' }
-    ],
-    default: Easings.Linear
-  },
-  fadeTimeField: {
-    type: 'number',
-    label: 'Fade time (ms)',
-    id: 'fadeTime',
-    min: 100,
-    max: 60000,
-    default: 2000
-  },
-  faderChangeField: {
-    type: 'number',
-    label: 'Change value (dB)',
-    id: 'value',
-    min: -100,
-    max: 100,
-    default: 3
-  },
-  volumeBusesDropdown: {
-    type: 'dropdown',
-    label: 'Bus',
-    id: 'bus',
-    choices: [
-      { id: 'solo', label: 'SOLO' },
-      { id: 'hp1', label: 'Headphone 1' },
-      { id: 'hp2', label: 'Headphone 2' }
-    ],
-    default: 'solo'
-  }
-};
+
+type SomeOptionField =
+	| CompanionInputFieldStaticText
+	| CompanionInputFieldColor
+	| CompanionInputFieldTextInput
+	| CompanionInputFieldDropdown
+	| CompanionInputFieldMultiDropdown
+	| CompanionInputFieldNumber
+	| CompanionInputFieldCheckbox
```

Comment:
Is there a built-in supertype for option fields that works for both feedbacks and actions? `SomeCompanionFeedbackInputField` would work but it has `feedback` in its name which I wanted to avoid.

Reply:
There isn't because they both support a slightly different list. For now actions is an extension of feedbacks, but I expect that at some point something will be added for just feedbacks.

I would be tempted to not give `OPTIONS` a type and suppress eslint. That way at the usage site, it will know exactly which one of the types it is

Reply:
Auto completion within this file was useful, but you're right and I removed the type.

Reply:
The side effect of this is that I'll have to add `as const` to each `type` property in options since wide `string` is not compatible to the type of `SomeCompanionFeedbackInputField` …
I guess I'll move it back to an explicit type.

Reply:
I use the `literal` helper exported from `@companion-mdoule/base` to enforce typings of each element individually.
eg https://github.com/bitfocus/companion-module-bmd-atem/blob/master/src/feedback.ts#L122